### PR TITLE
desc, select and insert into support for VARCHAR in Hive

### PR DESCRIPTION
**Pull request by losipiuk**
Replacement for https://github.com/TeradataCenterForHadoop/presto-td/pull/3

### CODE REVIEW

**Comment by pnowojski** [comment]
Hi, just checking what's the status of this pull request?

**Comment by sopel39** [comment]
looks good, but I don't like the fact that we loose loose varchar length information since it is now supported by @dain patch

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveUtil.java [hunk 1]
```diff
@@ -449,6 +453,11 @@ public static boolean isStructuralType(HiveType hiveType)
         return hiveType.getCategory() == Category.LIST || hiveType.getCategory() == Category.MAP || hiveType.getCategory() == Category.STRUCT;
     }
 
+    public static boolean isVarcharType(HiveType hiveType)
```
Comment (sopel39) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveUtil.java]: move above isArrayType

**Comment by kokosing** [comment]
I rebased and squashed commits.

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHiveFileFormats.java [hunk 2]
```diff
@@ -144,10 +146,12 @@ public void testRcTextPageSource()
     public void testRCBinary()
             throws Exception
     {
-        List<TestColumn> testColumns = ImmutableList.copyOf(filter(TEST_COLUMNS, testColumn -> {
-            // RC file does not support complex type as key of a map
-            return !testColumn.getName().equals("t_map_null_key_complex_key_value");
-        }));
+        // RC file does not support complex type as key of a map and interprets empty VARCHAR as nulls
+        List<TestColumn> testColumns = TEST_COLUMNS.stream()
+                .filter(testColumn -> {
```
Comment (haozhun) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestHiveFileFormats.java]: You can do `.map(TestColumn::getName)` followed by `filter`

**Comment by haozhun** [comment]
@losipiuk Looking at the pull request, I understand there are some file formats has problems dealing with `VARCHAR` and `CHAR`:

* If you write empty VARCHAR(50) in RCBinary format, and read it back. It will become null.
* If you write empty CHAR(50) in Parquet format, and read it back. It will still be empty.

Is there any other problems you observed? Are they a bug in writing, a bug in reading, or a deficiency in the file format itself?

File: presto-hive/src/test/sql/create-test-hive12.sql [hunk 3]
```diff
@@ -9,6 +9,7 @@ CREATE TABLE presto_test_types_textfile (
 , t_boolean BOOLEAN
 , t_timestamp TIMESTAMP
 , t_binary BINARY
+, t_varchar VARCHAR(50)
```
Comment (haozhun) [inline presto-hive/src/test/sql/create-test-hive12.sql]: Also add a test with `CHAR`?

**Comment by pnowojski** [comment]
@haozhun we haven't found other problems. RC file format at least by default encodes empty strings as nulls, so there is no way to write and later read an empty string. I don't remember at the moment whether DWRF supports varchar or not, but since it's deprecated should we bother?

**Comment by electrum** [comment]
We won't use varchar with DWRF, so I don't think it matters either way.

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHiveFileFormats.java (hunk 2 above)
Comment (pnowojski) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestHiveFileFormats.java]: That would change the type of the List to List<String>

File: presto-hive/src/test/sql/create-test-hive12.sql [hunk 4]
```diff
@@ -29,6 +30,7 @@ SELECT
 , CASE n % 3 WHEN 0 THEN false WHEN 1 THEN true ELSE NULL END
 , CASE WHEN n % 17 = 0 THEN NULL ELSE '2011-05-06 07:08:09.1234567' END
 , CASE WHEN n % 23 = 0 THEN NULL ELSE CAST('test binary' AS BINARY) END
+, CASE n % 39 WHEN 0 THEN NULL WHEN 1 THEN '' ELSE 'test varchar' END
```
Comment (pnowojski) [inline presto-hive/src/test/sql/create-test-hive12.sql]: value 39 is already being used for this column in create-test-hive13.sql. 37 is used by some dates

File: presto-hive/src/test/sql/create-test-hive12.sql (hunk 3 above)
Comment (pnowojski) [inline presto-hive/src/test/sql/create-test-hive12.sql]: > Char datatype will be introduced in Hive 0.13.0 (HIVE-5191).

File: presto-hive/src/test/sql/create-test-hive13.sql [hunk 5]
```diff
@@ -83,6 +83,9 @@ SELECT * FROM presto_test_types_textfile
 
 
 -- Parquet fails when trying to use complex nested types.
+-- Parquet is missing TIMESTAMP and BINARY, and for some reason
```
Comment (pnowojski) [inline presto-hive/src/test/sql/create-test-hive13.sql]: fixed

**Comment by pnowojski** [comment]
I have addressed @haozhun comments and pushed changes

File: presto-hive/src/test/sql/create-test-hive12.sql (hunk 4 above)
Comment (haozhun) [inline presto-hive/src/test/sql/create-test-hive12.sql]: I see. It's okay then.

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHiveFileFormats.java (hunk 2 above)
Comment (haozhun) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestHiveFileFormats.java]: You are correct.

**Comment by haozhun** [comment]
@pnowojski: Thanks Piotr. Please expect delays because of Thanksgiving.

@electrum: @dain told me about complexities that may arise as a result of supporting CHAR. Specifically, how various string functions should treat the trailing spaces. He said you've thought through this. I'll let you decide whether we are sure we want to support CHAR right now. And if we do, whether we should remove trailing spaces upon read.

@electrum and @martint: @dain, @cberner, and I discussed the other day about our plan of dealing with `VARCHAR(20)` where there are more than 20 code points. @dain believe we shouldn't commit to support this. And @cberner believe we should truncate them now to prevent potentially changing the result in such cases in the future. What's your opinion, @electrum, @martint?

**Comment by losipiuk** [comment]
I rebased onto master and pushed new version which drops CHAR(x) support. We left support for VARCHAR(x) in Hive which is translated to VARCHAR in presto. 
Due to semantical complications with CHAR we would like to postpone support to it until type parametrization patches are in.

cc: @pnowojski

**Comment by losipiuk** [comment]
ping :)

**Comment by haozhun** [comment]
Sorry. It's my fault to have not looked at it in the past month. Given the current progress of reviews for supporting `VARCHAR(n)`, @martint would like to come back to this pull requests after those changes are in place.

**Comment by losipiuk** [comment]
@haozhun  VARCHAR(n) work is in. I rebased this PR on top of master.
Are we still waiting for DECIMAL?

**Comment by martint** [comment]
@dain, can you look at this one?

**Comment by dain** [comment]
yep

**Comment by dain** [comment]
The Hive readers should truncate `VARCHAR` columns to the correct length on read.

**Comment by arhimondr** [comment]
Superseeded by #4781

**Comment by arhimondr** [comment]
@dain 
> The Hive readers should truncate VARCHAR columns to the correct length on read.

Do you think it is better to truncate the values instead of throw an exception? Someone can create a table with the less varchar length that it is required and never notice that his values are truncated.

**Comment by ilfrin** [comment]
@dain regarding your comment "The Hive readers should truncate VARCHAR columns to the correct length on read."

I'm wondering how to approach this so we are consistent across other data types. 
For example for DECIMAL if we encounter a string that has a precision greater than 38 we will error out. 

Lets not forget that
* it may cover up an issue with the data that the client should know about
* the client could have bad assumptions about the data and when we do this he will never know that

Then again there's the usual argument in big data, that this protects against running into a unusual value in a single row that will blow up a super long running query.

**Comment by dain** [comment]
The issue with `VARCHAR` is that Hive allows users to change the type without rewriting all of the data, and instead the reader is supposed to adapt if possible.  For `VARCHAR`, the readers truncate to length.  For DECIMAL, I believe it is ok to fail since Hive now restricts DECIMAL to 38.

**Pull request closed without merging**